### PR TITLE
Code analysis without codetools

### PR DESCRIPTION
**Pull request by wlandau**
# Summary

This PR implements fully custom code analysis for implicit dependency detection. Reasons:

1. Increase speed.
2. Remove the dependency on `codetools`.

# Speed

The speed improvement is not as much as I had hoped, but it is an improvement. Maybe we could improve it even further with profiling.

## Benchmark

```r
library(drake)
library(microbenchmark)
library(storr)

plan <- function(n){
  plan <- drake_plan(target_1 = 1)
  for (i in seq_len(n - 1) + 1){
    target <- paste0("target_", i)
    dependencies <- paste0("target_", seq_len(i - 1))
    command <- paste0("max(", paste0(dependencies, collapse = ", "), ")")
    plan <- rbind(plan, data.frame(target = target, command = command))
  }
  plan
}

overhead <- function(config){
  make(config = config)
  clean(cache = config$cache)
}

benchmarks <- microbenchmark(
  config <- drake_config(
    plan(500),
    verbose = 0,
    cache = storr_rds(tempfile())
  ),
  times = 10
)

print(benchmarks)
```

## Before this PR

```r
#> Unit: seconds
#>                                                                           expr
#>  config <- drake_config(plan(500), verbose = 0, cache = storr_rds(tempfile()))
#>       min       lq     mean   median       uq      max neval
#>  10.91839 10.93819 11.14081 11.13745 11.22658 11.67665    10
```

## After

```r
#> Unit: seconds
#>                                                                           expr
#>  config <- drake_config(plan(500), verbose = 0, cache = storr_rds(tempfile()))
#>       min       lq     mean   median       uq      max neval
#>  7.809276 7.814909 7.867705 7.844215 7.917026 7.966349    10
```

# Quality of the code analysis

From inspecting the `codetools` source code, I feel I understand when and how to register global variables detected from code. I even borrowed `codetools`'s own tests, and nearly all of them pass. However, there is one final snag: [replacement functions](https://stackoverflow.com/questions/11563154/what-are-replacement-functions-in-r).

``` r
d <- data.frame(x = 1)
colnames(d) <- "y"
d
#>   y
#> 1 1
```

Here is a case in which `codetools` detects `f<-` but `drake` does not:

``` r
g <- function(x){
  f(x) <- 9
}
codetools::findGlobals(g)
#> [1] "<-"  "{"   "f<-"
drake:::analyze_code(g)
#> list()
```

# Related GitHub issues and pull requests

- Ref: #573 

# Checklist

- [x] I have read `drake`'s [code of conduct](https://github.com/ropensci/drake/blob/master/CONDUCT.md), and I agree to follow its rules.
- [x] I have listed any substantial changes in the [development news](https://github.com/ropensci/drake/blob/master/NEWS.md).
- [x] I have added [`testthat`](https://github.com/r-lib/testthat) unit tests to [`tests/testthat`](https://github.com/ropensci/drake/tree/master/tests/testthat) to confirm that any new features or functionality work correctly.
- [x] I have tested this pull request locally with `devtools::check()`
- [x] This pull request is ready for review.
- [x] I think this pull request is ready to merge.

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -16,7 +16,8 @@
 - Allow the `magrittr` dot symbol to appear in some commands sometimes.
 - Deprecate the `fetch_cache` argument in all functions 
 - Remove packages `DBI` and `RSQLite` from "Suggests".
-- Define a special `config$eval <- new.env(parent = config$envir)` for storing built targets and evaluating commands in the plan. Now, `make()` no longer modifies the user's environemnt. This move is a long-overdue step toward purity.
+- Define a special `config$eval <- new.env(parent = config$envir)` for storing built targets and evaluating commands in the plan. Now, `make()` no longer modifies the user's environment. This move is a long-overdue step toward purity.
+- Remove depencency on the `codetools` package.
```
Comment (bpbond) [inline NEWS.md]: "dependency"

Comment (wlandau) [reply NEWS.md]: Yup, thanks. Fixed in 0a3dd3b528e496c54b5d3a305c60266917783626.

**Comment by wlandau** [comment]
@bpbond yeah, this stuff is interesting. In [`codetools`](https://github.com/cran/codetools), @ltierney went really deep to capture the idiosyncrasies that make R such a flexible language. I learned a lot from diving into the [`codetools`](https://github.com/cran/codetools) source code and preparing this PR.

The other cool piece is seeing how much we can speed up the code analysis now that we have more control. After seeing [these results](https://github.com/ropensci/drake/issues/573#issuecomment-448211610), I plan to try recruiting [hash tables](https://github.com/ropensci/drake/blob/master/R/hash_table.R) in more steps of the [code analysis](https://github.com/ropensci/drake/blob/master/R/code_analysis.R).

**Comment by ltierney** [comment]
codetools was forked out of an early version of the byte code
compiler. The current compiler is maintained as a literate program in
the R sources; the woven version linked from
http://developer.r-project.org/TechDocs/ may give a more readable
account of the language issues that need to be considered.

On Tue, 18 Dec 2018, Will Landau wrote:

> 
> @bpbond yeah, this stuff is interesting. In codetools, @ltierney went really
> deep to capture the idiosyncrasies that make R such a flexible language. I
> learned a lot from diving into the codetools source code and preparing this
> PR.
> 
> The other cool piece is seeing how much we can speed up the code analysis
> now that we have more control. After seeing these results, I plan to try
> recruiting hash tables in more steps of the code analysis.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the
> thread.[ADvaVCwmMIdYKyMizu7DJOeMjeNiXKbsks5u6O9QgaJpZM4ZXRac.gif]
> 
> 
>

-- 
Luke Tierney
Ralph E. Wareham Professor of Mathematical Sciences
University of Iowa                  Phone:             319-335-3386
Department of Statistics and        Fax:               319-335-3017
    Actuarial Science
241 Schaeffer Hall                  email:   luke-tierney@uiowa.edu
Iowa City, IA 52242                 WWW:  http://www.stat.uiowa.edu

**Pull request merged**